### ISSUE TITLE
simplification of Excel date handling

### ISSUE THREAD

**Issue by epogrebnyak** [epogrebnyak/FinancePy#3]
@domokane - I wrote a simplification of date handling with respect to Excel dates, date shifts and days of the week, please see if fits your code style:

https://github.com/epogrebnyak/FinancePy/blob/bd104d5a16863bbbe976f7653b259f90bbb79e28/financepy/finutils/dates.py#L1-L109

It should save a lot of back and forth with intermediate index. One can build `FinDate(day, month, year)` on top of `Date`.

**Comment by domokane** [comment]
This is interesting. If it makes it faster or easier to maintain then I am happy to revisit the implementation of FinDate.

But I would rather not start changing fundamental things like FinDate right now when I already have a list of things to do.

If it makes sense, this can be done in a few weeks after the branch changes we discussed have been done by me or jointly with you. I would like to make these branch changes as soon as I can as some people are requesting new functionality and I don’t want to delay that.

If you want to help, can you add some unit test cases to the new branch of the project so I can see how they will work and can start building my own unit tests by copying yours, and can you add the documentation generator you discussed ?

I will be doing more refactoring tonight so will push a new version late.

D

**Comment by epogrebnyak** [comment]
Well, FinDate also does need a test, maybe I should write more tests based on asserts in these implementation, so that existing and new implementation do pass? 

Were you able to your new tests with pytest? Also I did not notice the dev branch in the parent repo - are you sure you pushed it?

**Comment by domokane** [comment]
I agree. Unit tests for FinDate is good. FinancePy needs them and I am happy to have your suggestions.

I didn’t add any new tests yet – I was refactoring and renaming – I was hoping you would do one that I can use as a template.

Could you add one unit test to financepy – and more if you have time – here is the new branch

https://github.com/domokane/FinancePy/tree/more_pep_8

I can then use that as a template for the other tests. That would be a big step forward !!

And later, if you can get sphinx to work that would be good too.

Regards
D

**Comment by epogrebnyak** [comment]
Strange I do not see the branches on mobile, but I can follow a direct
link.

Will add few tests and try get the running locally and remote.

ср, 3 мар. 2021 г., 0:06 domokane <notifications@github.com>:

**Comment by epogrebnyak** [comment]
> I agree. Unit tests for FinDate is good. FinancePy needs them and I am happy to have your suggestions. I didn’t add any new tests yet – I was refactoring and renaming – I was hoping you would do one that I can use as a template. Could you add one unit test to financepy – and more if you have time – here is the new branch https://github.com/domokane/FinancePy/tree/more_pep_8 I can then use that as a template for the other tests. That would be a big step forward !! 

Here is pull request https://github.com/domokane/FinancePy/pull/78 for tests (from another fork). 

Few comments:

- moved golden tests to own folder (this couses conflict on merge for some reason)
- you need to install package locally with `pip install -e .` in root
- after that you invoke `pytest`
- added several demo tests Test_FinDate.py and disentangled the rest from golder runner
- wherever a file get renamed with underscore like `Test_` it gets disovered by pytest

**Comment by domokane** [comment]
Hi Evgeny
I merged your changes. How do I amend the yaml script on github to get the automatic run to work. It is now broken.
Regards
D

**Comment by epogrebnyak** [comment]
Will fix soon, we have a special issue on this.

ср, 3 мар. 2021 г., 17:21 domokane <notifications@github.com>:

**Comment by epogrebnyak** [comment]
@domokane - please see https://github.com/domokane/FinancePy/pull/80 for tests

You seem to have started a renaming - do you plan to freeze at some point? 

Few things a bit inconsistent: 
- module named like `Frequency` - should be frequency lowercase
- also `import *` is antipattern (use in init.py) - it brings all the stuff from modules into top level namespace

Suggest you write more tests before renaming or do some freeze on renaming  before there are more tests. Few example tests on FinDate (now Date) pass, but keeping up with renaming is a challenging.

**Comment by domokane** [comment]
Hi – Yes I started renaming – mostly input variable names, internal variables and some classes. Every time I make a non-trivial change I run all my tests – takes a few minutes. If nothing breaks I continue. I have more work to do so there are some inconsistencies which I will tidy up. So Frequency will become frequency as you noted etc...
Tomorrow I will work on class methods starting with the bond products. When I am done and check this in please take a look and give me your comments. I can then continue re-naming the other product folders.
D

**Comment by epogrebnyak** [comment]
Strange currently Date did not work without changing other parts of the
code.

Do you want to keep the existing custom tests in the CI (Github Actions)?
What role is there for unit tests?

вс, 7 мар. 2021 г., 2:04 domokane <notifications@github.com>:

**Comment by domokane** [comment]
I do want to use unit tests. But right now my priority is to change the interface asap to be more pep8 so I can release a new version and then get back to finance releases. I am not sure I have the time to make all of these changes by myself in the time I would like.
I will then work on adapting the existing tests to unit tests over time. So I would like the CI to do both the old tests and the new unit tests until this process is complete.
D
PS I am using PyCharm for the refactoring. What problem is there with Date ? Everything works for me. Perhaps I need to push again – strange as I always check that the tests are passed before I do this.

**Comment by epogrebnyak** [comment]
There is a problem your current verification procedure is long and does not provide a yes/no answer if the package works (you log the errors and go on without resolution).

Like here  - the import is broken, because you renamed file to `Frequency.py` and it should be the `frequency.py`:
 
https://github.com/domokane/FinancePy/blob/f4afd44876d284f3cf5ab225087ef0b50da941e8/financepy/utils/__init__.py#L5

With a unit test - you get a quick response - the package is not importable, tests fail without running at collection of test. 
You can ree that in a test run: https://github.com/domokane/FinancePy/runs/2047716431

```
 financepy/utils/__init__.py:5: in <module>
    from .frequency import *
E   ModuleNotFoundError: No module named 'financepy.utils.frequency'
```

Where I wanted a test on Date to pass I needed to correct the imports of frequency - because it was not either run or not picked up by existing golden tests:  

- https://github.com/domokane/FinancePy/pull/80/commits/3c031e3b73ddceec042a7b70bf2abc9823375959
- https://github.com/domokane/FinancePy/pull/80/commits/cfe916856a6690b0c096559392c22b4ef30780eb

So in your dev branch you commi tcode that does not run, even though perhaps accidentally. With a unit test you would have had a protection against that. Common rule not to commit anything from local to remote, unless tests pass locally. The golden tests are too long for that and you get commits that break the package. This is not a super major disaster, can happen, but you kind of shifts the burden to find out what is going on and eats out time. 

I see you want to do the renaming, but "we'll write unit tests later" is not realy saving time. 

The strategy can be to of write just few unit tests, one for anything in each submodule(), and trying to see what the coverage is. 

Also need a resolution on golden tests: what folder they are in (in https://github.com/domokane/FinancePy/runs/2047716431 they stay in `tests`) and if they should run on github actions.

**Comment by epogrebnyak** [comment]
To go further with unit tests we need to decide on following:

This PR: 

- [ ] unit tests, runnable by pytest, live in `test_pytest` folder
- [ ] currently there are 12 tests for `dates.py`
- [ ] golden tests live in `tests` as they were
- [ ] pytest unit tests runs on Github Actions
- [ ] contributors and maintainer are asked to run pytest locally before commit
- [ ] pytest badge shown in readme

To decide:

- [x] golden tests do not run on Github Actions (this PR - because they just provide warnings and run long)
- [ ] golden tests do run on Github Actions (because they write back to golden folder and can be reviewed manually)

Short term:
- list public classes and methods that make package API
- add more unit tests (I can expand  a few tests I tsarted)
- show test coverage

Longer term:

- [ ] golden tests refactored in as a logging procedure for performance/numeric verification checks
    - there is a logger module, not part testing framework
    - logging runs/procedures, not part testing framework
    - control entity that takes care of current vs golden outputs, can be used as tests
- [ ] golden tests go to `tests_golden` folder
- [ ] pytest tests go into standard location `tests` folder

**Comment by domokane** [comment]
I told you via email that I am not sure I am going to be able to write 102 sets of unit tests before I release the next version but I or someone else will do it over time. Between now and then I want to keep running tests_golden and the unit tests.

**Comment by domokane** [comment]
I have just seen your latest comment. I am fine with what you are saying. We want the same thing. It's simply a difference in timing - I have a new and major work distraction right now that means that I cannot devote the time I would like to the unit tests. So my priority is to change the interface to be more pep8 compliant and to release this change asap. With your help I also wish to get a framework for units tests in place using pytest and to get a more standard documentation system in place.

**Comment by epogrebnyak** [comment]
So `tests_golden` for golden tests and `tests` for pytest?

**Comment by domokane** [comment]
DETAILED RESPONSE IN CAPS BELOW:

To go further with unit tests we need to decide on following:

This PR:

 unit tests, runnable by pytest, live in test_pytest folder - AS MANY UNIT TESTS AS POSSIBLE BUT 102 SETS NOT POSSIBLE
 currently there are 12 tests for dates.py - GOOD
 golden tests live in tests as they were - WE CAN KEEP THEM IN FOLDER tests_golden WHY RENAME AGAIN ?
 pytest unit tests runs on Github Actions SURE
 contributors and maintainer are asked to run pytest locally before commit I WOULD ASK THEM TO RUN BOTH UNTIL ALL UNIT TESTS HAVE BEEN DONE. IF THE COVERAGE IS INCOMPLETE THEN TESTS_GOLDEN MUST BE RUN.
 pytest badge shown in readme NOT FAMILIAR WITH THIS
To decide:

 golden tests do not run on Github Actions (this PR - because they just provide warnings and run long)  DO BOTH
 golden tests do run on Github Actions YES AND UNIT TESTS
Short term:

list public classes and methods that make package NOT SURE WHAT THIS IS FOR
add more unit tests (I can expand a few tests I tsarted) GREAT
show test coverage ?

Longer term:

 golden tests refactored in as a logging procedure for performance/numeric verification checks YEP. AND AS EXAMPLES OF HOW TO CALL FINANCEPY. AGREE 100%
 golden tests go to tests_golden folder YES - OR MAYBE A BETTER NAME CAN BE FOUND. BUT NOT tests
 pytest tests go into standard location tests folder YES

**Comment by domokane** [comment]
YES - pytests in tests. Old tests in tests_golden. Maybe I can find a better name as "golden" is not meaningful. But that can be decided later as it does not affect the API.

**Comment by epogrebnyak** [comment]
> YES - pytests in tests. Old tests in tests_golden. Maybe I can find a better name as "golden" is not meaningful. But that can be decided later as it does not affect the API.

Implemented in https://github.com/domokane/FinancePy/pull/80 - see if you can merge into dev branch. The golden tests run on CI and line in `tests_golden`.

**Comment by epogrebnyak** [comment]
You need to rename `Frequency.py` lowercase `frequency.py` too:

```
TEST:  90 out of 102: MODULE: TestFinNumbaNumpySpeed              Base error: No module named 'financepy.utils.frequency'
TEST:  91 out of 102: MODULE: TestFinOIS                          Base error: No module named 'financepy.utils.frequency'
TEST:  92 out of 102: MODULE: TestFinOISCurve                     Base error: No module named 'financepy.utils.frequency'
TEST:  93 out of 102: MODULE: TestFinOptionImpliedDbn             Base error: No module named 'financepy.utils.frequency'
TEST:  94 out of 102: MODULE: TestFinPieceCurve                   WARNINGS:   0   ERRORS:   0 
TEST:  95 out of 102: MODULE: TestFinProcessSimulator             WARNINGS:   0   ERRORS:   0 
TEST:  96 out of 102: MODULE: TestFinSchedule                     Base error: No module named 'financepy.utils.frequency'
TEST:  97 out of 102: MODULE: TestFinSobol                        WARNINGS:   2   ERRORS:   0 
TEST:  98 out of 102: MODULE: TestFinStatistics                   WARNINGS:   3   ERRORS:   0 
TEST:  99 out of 102: MODULE: TestFinSwapLegs                     Base error: No module named 'financepy.utils.frequency'
TEST: 100 out of 102: MODULE: TestFinSwaptionVolSurface           Base error: No module named 'financepy.utils.frequency'
```

**Comment by domokane** [comment]
I have done it. Will push. Tell me - what do you do about removing .pyc files ? Is this something you do automatically before running tests ?

**Comment by domokane** [comment]
> > YES - pytests in tests. Old tests in tests_golden. Maybe I can find a better name as "golden" is not meaningful. But that can be decided later as it does not affect the API.
> 
> Implemented in [domokane#80](https://github.com/domokane/FinancePy/pull/80) - see if you can merge into dev branch. The golden tests run on CI and line in `tests_golden`.

I already made the change too before I saw this. If it's different/better than what I did I will merge.

**Comment by epogrebnyak** [comment]
The folder naming should be the same  the same as yours, but a few changes
in test runner. Try pull and merge my branch,  I suggest

вс, 7 мар. 2021 г., 13:26 domokane <notifications@github.com>:

**Comment by epogrebnyak** [comment]
At what point are .pyc files cause a  problem? We should have them in
gitignore, so they are created locally, but do not make ot into remote
reposotiry.

Are they polluting an output of golden tests (until we come up with a
better name)?

There is usually no special treatment of pyc  files, other than gitignore

вс, 7 мар. 2021 г., 13:24 domokane <notifications@github.com>:

**Comment by domokane** [comment]
The issue is that when I run my test script it appears that it does not always refresh all the .pyc files if the .py files change. So the output from running the tests may not always be the same as the actual .py files. When I then push to the repository, it may not work in CI as it had to rebuild the .pyc files. I may be incorrect but I have had issues like this a few times.

**Comment by domokane** [comment]
I looked at your pull request. It reverses a lot of the naming refactoring I have done. I am not sure how to safely merge this without losing my renaming changes. I will do a push now. Can you pull this version and make the changes to that version ?

**Comment by epogrebnyak** [comment]
The pyc files do not run on their own, they are created when py is invoked,
in usual circumstances. (Unless something is intercepted)

We should delete any pyc files from repo and put the in gitignore, this way
there are no pyc files at all on a remote machine, when tests are run..

вс, 7 мар. 2021 г., 13:58 domokane <notifications@github.com>:
